xcm/src/v3/mod.rs

```diff
@@ -0,0 +1,1172 @@
+		self.0.last()
+	}
+
+	/// Return the only instruction, contained in `Self`, iff only one exists (`None` otherwise).
+	pub fn only(&self) -> Option<&Instruction<Call>> {
+		if self.0.len() == 1 {
+			self.0.first()
+		} else {
+			None
+		}
+	}
+
+	/// Return the only instruction, contained in `Self`, iff only one exists (returns `self`
+	/// otherwise).
+	pub fn into_only(mut self) -> core::result::Result<Instruction<Call>, Self> {
+		if self.0.len() == 1 {
+			self.0.pop().ok_or(self)
+		} else {
+			Err(self)
+		}
+	}
+}
+
+impl<Call> From<Vec<Instruction<Call>>> for Xcm<Call> {
+	fn from(c: Vec<Instruction<Call>>) -> Self {
+		Self(c)
+	}
+}
+
+impl<Call> From<Xcm<Call>> for Vec<Instruction<Call>> {
+	fn from(c: Xcm<Call>) -> Self {
+		c.0
+	}
+}
+
+/// A prelude for importing all types typically used when interacting with XCM messages.
+pub mod prelude {
+	mod contents {
+		pub use super::super::{
+			send_xcm, validate_send, Ancestor, AncestorThen,
+			AssetId::{self, *},
+			AssetInstance::{self, *},
+			BodyId, BodyPart, Error as XcmError, ExecuteXcm,
+			Fungibility::{self, *},
+			Instruction::*,
+			InteriorMultiLocation,
+			Junction::{self, *},
+			Junctions::{self, *},
+			MaybeErrorCode, MultiAsset,
+			MultiAssetFilter::{self, *},
+			MultiAssets, MultiLocation,
+			NetworkId::{self, *},
+			OriginKind, Outcome, PalletInfo, Parent, ParentThen, PreparedMessage, QueryId,
+			QueryResponseInfo, Response, Result as XcmResult, SendError, SendResult, SendXcm,
+			Unwrappable,
+			WeightLimit::{self, *},
+			WildFungibility::{self, Fungible as WildFungible, NonFungible as WildNonFungible},
+			WildMultiAsset::{self, *},
+			XcmWeightInfo, VERSION as XCM_VERSION,
+		};
+	}
+	pub use super::{Instruction, Xcm};
+	pub use contents::*;
+	pub mod opaque {
+		pub use super::{
+			super::opaque::{Instruction, Xcm},
+			contents::*,
+		};
+	}
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, TypeInfo)]
+pub struct PalletInfo {
+	#[codec(compact)]
+	pub index: u32,
+	pub name: Vec<u8>,
+	pub module_name: Vec<u8>,
+	#[codec(compact)]
+	pub major: u32,
+	#[codec(compact)]
+	pub minor: u32,
+	#[codec(compact)]
+	pub patch: u32,
+}
+
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, TypeInfo)]
+pub enum MaybeErrorCode {
+	Success,
+	Error(Vec<u8>),
+}
+
+impl Default for MaybeErrorCode {
+	fn default() -> MaybeErrorCode {
+		MaybeErrorCode::Success
+	}
+}
+
+/// Response data to a query.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, TypeInfo)]
+pub enum Response {
+	/// No response. Serves as a neutral default.
+	Null,
+	/// Some assets.
+	Assets(MultiAssets),
+	/// The outcome of an XCM instruction.
+	ExecutionResult(Option<(u32, Error)>),
+	/// An XCM version.
+	Version(super::Version),
+	/// The index, instance name, pallet name and version of some pallets.
+	PalletsInfo(Vec<PalletInfo>),
+	/// The status of a dispatch attempt using `Transact`.
+	DispatchResult(MaybeErrorCode),
+}
```

Comment:
Feature Request (maybe for v4):
It would be great to be able to generically query data. Either via `Transact` calls or a dedicated `QueryData` instruction (or whatever naming) and then get a `QueryResponse` with a blob of data.
This would avoid the overhead (mostly in terms of the queried party needing to know the query recipients call encoding) associated with `Transact` and allow defining custom protocols on top of XCM.
E.g. a use case I'm playing around with at the moment would require querying DID (decentralized identity) information over XCM. Currently the only way of implementing this that I see requires both chains to know each other's Call encoding (to be able to send `Transact`s), even though it should be easy to register a query response handler that expects a certain blob of data (that confirms or disconfirms the identity).

Reply:
How would the data be referenced? Maybe submit a PR to the xcm-format repo?